ignite/engine/__init__.py

```diff
@@ -118,68 +320,31 @@ def create_supervised_trainer(
 
-        - Added ``amp`` argument for automatic mixed precision.
+        - Added ``amp_mode`` argument for automatic mixed precision.
         - Added ``scaler`` argument for GradScaler instance.
-        - Added ``grad_norm_kwargs`` argument for :func:`~torch.nn.utils.clip_grad_norm_`.
     """
 
     device_type = device.type if isinstance(device, torch.device) else device
     on_tpu = "xla" in device_type if device_type is not None else False
-    _arg_check(on_tpu, amp, scaler)
-    has_native_amp, has_apex_amp = False, False
-
-    if amp and LooseVersion(torch.__version__) >= LooseVersion("1.6.0") and not on_tpu:
-        from torch.cuda.amp import autocast
-
-        has_native_amp = True
-    elif amp and not on_tpu:
-        try:
-            from apex import amp as apex_amp
-
-            has_apex_amp = True
-        except ImportError:
-            raise ImportError("Please install apex from https://github.com/nvidia/apex to use amp.")
-
-    def _update(engine: Engine, batch: Sequence[torch.Tensor]) -> Union[Any, Tuple[torch.Tensor]]:
-        model.train()
-        optimizer.zero_grad()
-        x, y = prepare_batch(batch, device=device, non_blocking=non_blocking)
-        if has_native_amp:
-            with autocast(enabled=amp):
-                y_pred = model(x)
-                loss = loss_fn(y_pred, y)
-        else:
-            y_pred = model(x)
-            loss = loss_fn(y_pred, y)
-
-        if has_native_amp and scaler is not None:
-            # scaled the loss
-            scaler.scale(loss).backward()
-        elif has_apex_amp:
-            with apex_amp.scale_loss(loss, optimizer) as scaled_loss:
-                scaled_loss.backward()
-        else:
-            loss.backward()
-
-        if grad_norm_kwargs and has_apex_amp:
-            torch.nn.utils.clip_grad_norm_(apex_amp.master_params(optimizer), **grad_norm_kwargs)
-        elif grad_norm_kwargs:
-            if has_native_amp and scaler is not None:
-                # unscale the optimizer for clip_grad_norm_
-                scaler.unscale_(optimizer)
-            # clip grad norm
-            torch.nn.utils.clip_grad_norm_(model.parameters(), **grad_norm_kwargs)
-
-        if on_tpu:
-            xm.optimizer_step(optimizer, barrier=True)
-        elif has_native_amp and scaler is not None:
-            # unscale_() is called inside step()
-            scaler.step(optimizer)
-            scaler.update()
-        else:
-            optimizer.step()
-
-        return output_transform(x, y, y_pred, loss)
+    _arg_check(on_tpu, amp_mode, scaler)
+
+    if amp_mode == "amp":
+        _update = supervised_trainer_step_amp(
+            model, optimizer, loss_fn, device, non_blocking, prepare_batch, output_transform, scaler
+        )
+    elif amp_mode == "apex":
+        _update = supervised_trainer_step_apex(
+            model, optimizer, loss_fn, device, non_blocking, prepare_batch, output_transform
+        )
+
+    if on_tpu:
```

Comment:
```suggestion
    elif on_tpu:
```
Function `_check_arg` ensures that all parameters is coherent if I understand correctly. So 
```
flag = _check_arg(...)

# no need to understand the rules in _check_arg
# apply the relevant mode
if flag == "amp":
   ...
elif flag == "apex":
   ...
elif ...
```

Reply:
Thanks @sdesrozis